File: zig/collatz-conjecture/collatz_conjecture.zig
pub const ComputationError = error {IllegalArgument};

pub fn steps(number: isize) anyerror!usize {
    if (number < 1) {
        return ComputationError.IllegalArgument;
    } else {
        var stepCount: usize = 0;
        var n: isize = number;
        while (n > 1) {
            n = if (@rem(n, 2) == 0) @divTrunc(n, 2) else (3 * n + 1);
            stepCount += 1;
        }
        return stepCount;
    }
}

File: ruby/collatz-conjecture/collatz_conjecture.rb
class CollatzConjecture
  def self.steps(number)
    number > 0 or raise ArgumentError
    count = 0
    while number != 1
      number = number.even? ? number / 2 : 3 * number + 1
      count += 1
    end
    count
  end
end
